docs/reference/isr_rules.rst

```diff
@@ -0,0 +1,270 @@
+language. Others are common to all systems capable of responding to real time events. This document covers
+the language specific issues first, followed by a brief introduction to real time programming for those new to it.
+
+This introduction uses vague terms like "slow" or "as fast as possible". This is deliberate, as speeds are
+application dependent. Acceptable durations for an ISR are dependent on the rate at which interrupts occur,
+the nature of the main program, and the presence of other concurrent events.
+
+MicroPython Issues
+------------------
+
+The emergency exception buffer
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+If an error occurs in an ISR, MicroPython is unable to produce an error report unless a special buffer is created
+for the purpose. Debugging is simplified if the following code is included in any program using interrupts.
+
+.. code:: python
+
+    import micropython
+    micropython.alloc_emergency_exception_buf(100)
+
+Simplicity
+~~~~~~~~~~
+
+For a variety of reasons it is important to keep ISR code as short and simple as possible. It should do only what
+has to be done immediately after the event which caused it: operations which can be deferred should be delegated
+to the main program loop. Typically an ISR will deal with the hardware device which caused the interrupt, making
+it ready for the next interrupt to occur. It will communicate with the main loop by updating shared data to indicate
+that the interrupt has occurred, and it will return. An ISR should return control to the main loop as quickly
+as possible. This is not a specific MicroPython issue so is covered in more detail :ref:`below <ISR>`.
+
+Communication between an ISR and the main program
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Normally an ISR needs to communicate with the main program. The simplest means of doing this is via one or more
+shared data objects, either declared as global or shared via a class (see below). There are various restrictions
+and hazards around doing this, which are covered in more detail below. Integers, ``bytes`` and ``bytearray`` objects
+are commonly used for this purpose along with arrays (from the array module) which can store various data types.
+
+The use of object methods as callbacks
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+MicroPython supports this powerful technique which enables an ISR to share instance variables with the underlying
+code. It also enables a class implementing a device driver to support multiple device instances. The following
+example causes two LED's to flash at different rates.
+
+.. code:: python
+
+    import pyb, micropython
+    micropython.alloc_emergency_exception_buf(100)
+    class Foo(object):
+        def __init__(self, timer, led):
+            self.led = led
+            timer.callback(self.cb)
+        def cb(self, tim):
+            self.led.toggle()
+
+    red = Foo(pyb.Timer(4, freq=1), pyb.LED(1))
+    greeen = Foo(pyb.Timer(2, freq=0.8), pyb.LED(2))
```

Comment:
I think there needs to be a paragraph here which describes the code above, what the interrupts are and what they call.  Would also be good to explain that self.cb is a bound method which is how different state is passed around.